### ISSUE TITLE
Structured logging with lists & dicts

### ISSUE THREAD

**Issue by eitel13** [Delgan/loguru#1040]
Hey @Delgan ! 👋 🙂 

I've been through quite some "issues" with a similar topic (including https://github.com/Delgan/loguru/issues/424 which I have also participated in! ) but I thought perhaps a new thread is worth it, instead of reviving a closed one. 😅

# Logging `Lists` & `Dicts`

Previously, you helped me get to [this](https://github.com/Delgan/loguru/issues/424#issuecomment-1068461212) point:

```Python
import sys

from loguru import logger


def formatter(record):
    base_format = "{time} {level} {name} {message} " + " " * 10
    base = base_format.format_map(record)
    lines = str(record["extra"].get("data", "")).splitlines()
    indent = "\n" + " " * len(base)
    reformatted = base + indent.join(lines)
    record["extra"]["reformatted"] = reformatted
    return "{extra[reformatted]}\n{exception}"


logger.remove()
logger.add(sys.stderr, format=formatter)


data = """---------------- Request ----------------
Headers       :   {"Accept": "*/*",
                   "Accept-Encoding": "gzip, deflate",
                   "Connection": "keep-alive",
                   "Content-Length": "20",
                   "User-Agent": "python-requests/2.27.1",
                   "cookie": "foo=bar; bar=baz",
                   "x-pretty-print": "2"}
URL           :   http://mockbin.com/request/test/1/mock?foo=bar&foo=baz
Method        :   POST
Body          :   {"foo": "bar"}"""

logger.info("Default message")
logger.bind(data=data).info("Message with data")
```

However, in the above example, I was logging a `str`, and instead, I'd like to log a `dict`:

```JSON
{
        "Headers": {
            "Accept": "*/*",
            "Accept-Encoding": "gzip, deflate",
            "Connection": "keep-alive",
            "Content-Length": "20",
            "User-Agent": "python-requests/2.27.1",
            "cookie": "foo=bar; bar=baz",
            "x-pretty-print": "2",
        },
        "URL": "http://mockbin.com/request/test/1/mock?foo=bar&foo=baz",
        "Method": "POST",
        "Body": {
            "foo": "bar",
        },
    }
```

1. Is there a newer, _better_ way of the above solution? Have things changed perhaps? 👀 
2. How could I make it so that I can "auto-format" the output for `lists` or `dicts`?

# Logging with indents

Wrt OP's original query, I'd like to also have indentation, but slightly different.

This was the proposed solution:

```Python
logger_indentation = ContextVar("logger_indentation", default=0)


@contextmanager
def indent_logs(indent_size):
    val = logger_indentation.get()
    logger_indentation.set(val + indent_size)
    yield
    logger_indentation.set(val)


def patcher(record):
    indentation = logger_indentation.get()
    record.update(message=" " * indentation + record["message"])
```

And in my use-case, I am doing this:

```Python
  logger.info(f"Parameters:")
    for param_info in param_info_list:
      with indent_logs(4):
        logger.info(param_info)
```

With output:

```shell
2023-12-04 15:31:27.057 | INFO     | actions_wrapper:action_wrapper:129 - Parameters:
2023-12-04 15:31:27.057 | INFO     | actions_wrapper:action_wrapper:132 -     element = "some_str"
2023-12-04 15:31:27.058 | INFO     | actions_wrapper:action_wrapper:132 -     element_index = 0 (Default)
```

But, what I'd preferably like to have is one log statement:

```shell
2023-12-04 15:31:27.057 | INFO     | actions_wrapper:action_wrapper:129 - Parameters:
                                                                             element         = "some_str"
                                                                             element_index   = 0 (Default)
```
                        
And preferably not having to use a context manager - maybe something similar to using `bind()`? 🤔

**Comment by Delgan** [comment]
Hi @eitel13. :)

> Is there a newer, better way of the above solution? Have things changed perhaps? 👀

This is still the preferred method. First, you can attach contextual data using `bind()`. Secondly, you can format them according to your preferences using a custom `format` or `patcher`.


> How could I make it so that I can "auto-format" the output for `lists` or `dicts`?

I'm not sure what you mean exactly. Surely you can write a function that transform a `dict` to a formatted string, right? Basically, it could be `repr(data)`. You can also take a look at the [`pprint`](https://docs.python.org/3/library/pprint.html) module.



> But, what I'd preferably like to have is one log statement:
> 
> ```shell
> 2023-12-04 15:31:27.057 | INFO     | actions_wrapper:action_wrapper:129 - Parameters:
>                                                                              element         = "some_str"
>                                                                              element_index   = 0 (Default)
> ```
> And preferably not having to use a context manager - maybe something similar to using `bind()`? 🤔

It's complicated, because the size of the elements preceding can't be known in advance, generally.

You can either indent them with a fixed size:

```python
from loguru import logger

def indent_params(params):
    return "".join("\n" + " "  * 10 + str(key) + " = " + str(value) for key, value in params.items())

params = {"element": "some_str", "element_index": 0}
logger.info("Parameters: {}", indent_params(params))
```

Or you would have to pre-format the message by yourself:

```python
from loguru import logger
import sys

def custom_formatter(record):
    my_format = "{time} | {level} | {message}"
    my_message = my_format.format_map(record)
    indentation = len(my_message)
    if "params" in record["extra"]:
        items = record["extra"]["params"].items()
        params = "".join(f"{indentation * ' '} {key!s} = {value!s}\n" for key, value in items)
        my_message += "\n" + params
    record["extra"]["my_message"] = my_message
    return "{extra[my_message]}"

logger.add(sys.stderr, format=custom_formatter)

params = {"element": "some_str", "element_index": 0}
logger.info("Parameters: ", params=params)

```

**Comment by eitel13** [comment]
Hey @Delgan, thank you for such a quick reply! 💪 

> This is still the preferred method.

Awesome! Just making sure 🤓 

> I'm not sure what you mean exactly. 

Apologies 😅 I guess what I was suggesting is if there could be some internal mechanism of Loguru to handle logging of lists or dicts "natively" perhaps? Assuming you go with the `logger.bind()` approach.. 

Perhaps there could be some kind of an object type (a `dataclass`?) that has defined parameters, for example:

```Python
@dataclass
class PprintDict:
    """Indent dictionary class."""

    dictionary: OrderedDict
    indent: int = field(default=0)


@dataclass
class PprintList:
    """Indent list class."""

    list: list
    indent: int = field(default=0)
```

And you simply pass this object: 

```Python
logger.bind(data=IndentDict(indent=4, dictionary=some_dict))
logger.bind(data=IndentList(indent=2, dictionary=some_list))
```

Of course I do I know this is doable on an individual basis with a custom formatting approach ☝️ 

If I had to guess, probably many users would enjoy such a functionality without having to "handle" it on their end. 🙂 

> You can also take a look at the [pprint](https://docs.python.org/3/library/pprint.html) module.

I have used this module before (for simple printing) but I wasn't aware (IIUC) that it can propagate it's output to Loguru.. Is this the case? Because then my suggestion is pointless 🤣  

> because the size of the elements preceding can't be known in advance

Fair point! 👍 

Thank you so much @Delgan! I still think you've got an amazing package and I find your responses super helpful 💪

**Comment by eitel13** [comment]
In the meantime 😄 I've modified the formatter function to this:

```Python
def formatter(record):
    def_format = f"\n{{time:DD/MMM/YYYY H:M:S}} - [{{level}}] {{message}}" + " " * 3
    base = def_format.format_map(record)
    data = record["extra"].get("data", "")

    if isinstance(data, dict):
        data_str = json.dumps(data, indent=4)
        lines = [line.rstrip() for line in data_str.splitlines()]
    else:
        lines = [str(data)]

    indent = "\n" + (" " * (len(base) - 1))
    reformatted = base + indent.join(lines)

    record["extra"]["reformatted"] = reformatted
    return "<green>{extra[reformatted]}</green>\n{exception}"
```

Usage like this:

```Python
data = {
    "Headers": {
        "Accept": "*/*",
        "Accept-Encoding": "gzip, deflate",
        "fields": [
            "foo",
            "bar",
        ],
    },
    "URL": "http://mockbin.com/request/test/1/mock?foo=bar&foo=baz",
    "Method": "POST",
    "Body": {
        "foo": "bar",
    },
}

# Set up the logger with the custom formatter method
logger.add(sys.stderr, format=formatter, level="DEBUG")
logger.bind(data=data).info("User information:")
````
Output:

```shell
05/Dec/2023 11:12:6 - [INFO] User information:   {
                                                    "Headers": {
                                                        "Accept": "*/*",
                                                        "Accept-Encoding": "gzip, deflate",
                                                        "fields": [
                                                            "foo",
                                                            "bar"
                                                        ]
                                                    },
                                                    "URL": "http://mockbin.com/request/test/1/mock?foo=bar&foo=baz",
                                                    "Method": "POST",
                                                    "Body": {
                                                        "foo": "bar"
                                                    }
                                                }
```

I don't know how practical it would be to build such a "handler" into Loguru such that we don't need to have custom formatter functions, but it would be a welcomed addition! 😄 (especially if we intend to stick with the default format).

One side note, I do want to stick to the default logging format, but if I do this:

```Python
def formatter(record):
    def_format = LOGURU_FORMAT

    # rest of code....
```

Then the output log is like this:

```shell
<green>2023-12-05 11:39:04.773</green> | <level>INFO    </level> | <cyan>__main__</cyan>:<cyan><module></cyan>:<cyan>55</cyan> - <level>User information:</level>
```

**Comment by eitel13** [comment]
Since the only goal is to handle whatever is passed to the `bind()` method, is it possible to "add" this functionality while not affecting the configured logger? (Perhaps something similar to https://github.com/Delgan/loguru/issues/239)

As I understand, you:

1. Define your log format (which I do in the function `def formatter`)
2. Remove the current formatter (to avoid duplicates in the log output)
3. Add your custom formatter

From an example in the [`patch()`](https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.patch) section of the documentation, there's this:

```Python
# Set a default "extra" dict to logger across all modules, without "bind()"
extra = {"context": "foo"}
logger.configure(extra=extra)
logger.add(sys.stderr, format="{extra[context]} - {message}")

logger.info("Context without bind")
# => "foo - Context without bind"

logger.bind(context="bar").info("Suppress global context")
# => "bar - Suppress global context"
```

But this outputs 2 sets of logs:

```Shell
2023-12-05 15:23:50.455 | INFO     | __main__:<module>:14 - Context without bind
foo - Context without bind

2023-12-05 15:23:50.456 | INFO     | __main__:<module>:16 - Suppress global context
bar - Suppress global context
```

Which is to be expected because of `logger.add`.

But what if instead, all that was desired was:

```Shell
2023-12-05 15:23:50.455 | INFO     | __main__:<module>:14 - foo - Context without bind
2023-12-05 15:23:50.456 | INFO     | __main__:<module>:16 - bar - Suppress global context
```
In my case, I of course want `foo` & `bar` _after_ the message:

```Shell
2023-12-05 15:23:50.455 | INFO     | __main__:<module>:14 - Context without bind - foo
2023-12-05 15:23:50.456 | INFO     | __main__:<module>:16 - Suppress global context - bar
```
Basically, what I think I'm trying to get at is some way to format / influence the output of whatever is provided to the `bind()` method without overwriting the entire formatter. 😅

**Comment by Delgan** [comment]
Hey @eitel13. Sorry for the late answer.


> I guess what I was suggesting is if there could be some internal mechanism of Loguru to handle logging of lists or dicts "natively" perhaps?

I  think that is out-of-scope for Loguru. There may be different possible representations for the same object (using `str()`, `repr()`, `pprint()`, etc.), and each user may have their own preferences. Loguru opted for the default Python object representation that most of us are familiar with, namely `str()`. Advanced use cases like yours are rare I think, and I prefer to leave it up to the user to choose the appropriate representation rather than trying to provide a customized one.


> I have used this module before (for simple printing) but I wasn't aware (IIUC) that it can propagate it's output to Loguru.. Is this the case?

One could use [`PrettyPrinter.pformat()`](https://docs.python.org/3/library/pprint.html#pprint.pformat) to produce a `str` that can be logged as an argument of `logger.info()` or used in a custom `format` function.


> I don't know how practical it would be to build such a "handler" into Loguru such that we don't need to have custom formatter functions, but it would be a welcomed addition! 😄 (especially if we intend to stick with the default format).

Yeah, sorry, but as explained above, I think it's too much of a _niche_ for me to integrate it as a built-in. :sweat_smile:


> One side note, I do want to stick to the default logging format, but if I do this:
> 
> ```python
> def formatter(record):
>     def_format = LOGURU_FORMAT
> 
>     # rest of code....
> ```
> Then the output log is like this:
>
> ```
><green>2023-12-05 11:39:04.773</green> | <level>INFO    </level> | <cyan>__main__</cyan>:<cyan><module></cyan>:<cyan>55</cyan> - <level>User information:</level>
>```

This is because the color markups end up in the `extra["reformatted"]` string but Loguru does not recursively colorize the message (to avoid unexpected errors when unrelated tags are logged). The color markups need to be part of the string returned by the formatter to be taken into account:

```python

def formatter(record) -> str:
    from loguru._defaults import LOGURU_FORMAT

    def_format = (
        "{time:YYYY-MM-DD HH:mm:ss.SSS} | "
        "{level: <8} | "
        "{name}:"
        "{function}:"
        "{line} - "
        "{message}"
    )

    base = def_format.format_map(record)
    data = record["extra"].get("data", "")

    if isinstance(data, dict):
        data_str = json.dumps(data, indent=4)
        lines = [line.rstrip() for line in data_str.splitlines()]
        lines.insert(0, "")
    elif isinstance(data, list):
        lines = [f"{item}" for item in data]
        lines.insert(0, "")
    else:
        lines = [str(data)]

    indent = "\n" + (" " * (len(base.replace(record["message"], "").strip()) + 1))

    record["extra"]["formatted_data"] = indent.join(lines)
    return LOGURU_FORMAT + "{extra[formatted_data]}\n{exception}"
```


> Basically, what I think I'm trying to get at is some way to format / influence the output of whatever is provided to the `bind()` method without overwriting the entire formatter. 😅

The `extra` dict is not part of the default `format`, so you have no choice but to configure the `logger` with a new handler (and remove the default one to avoid duplicated logs).

**Comment by eitel13** [comment]
Happy new year @Delgan !  🎆  Another year to say thanks for this package! 🤣 

I appreciate the reply. I was considering closing because (as you mentioned below) I gathered it's more a user-issue than Loguru one. 🙇 

> Advanced use cases like yours are rare I think

Yeah, I figured too! I understand 🙂 

> One could use [PrettyPrinter.pformat()](https://docs.python.org/3/library/pprint.html#pprint.pformat) to produce a str

I will have a look, thank you!

> This is because the color markups end up in the extra["reformatted"] string... The color markups need to be part of the string returned by the formatter

Thank you _again_ 🤣 

There's a lot of nuances with Python logging (which I am veeeery unfamiliar with) and that's also why I love this package, because for the most part you've managed to strip away a lot of other complexities. I appreciate you bearing with me to explain these things 💪 

Is this `return LOGURU_FORMAT + "{extra[formatted_data]}\n{exception}"` something that could be (or maybe it is already 👀 ) documented somewhere?

It wasn't necessarily obvious to do this - I'm thinking perhaps for others too 🤷 

> so you have no choice but to configure the logger with a new handler

Ahh okay! The reason I was suggesting that is for an example like below 👇 

Imagine we have this:

- project A --> support module
- project B --> has a dependency on project A

In project A, we create a custom formatter (say to format `dicts` in this case), but we only intend to log this custom format of `dicts` within project A.

Wouldn't importing project A into project B also affect the logging of dicts contained within project B? 

If so, I think that would be undesirable behaviour? 🤔

**Comment by Delgan** [comment]
> Happy new year @Delgan ! 🎆 Another year to say thanks for this package! 🤣

Thank you @eitel13! I wish you a happy new year as well. :)


> Is this `return LOGURU_FORMAT + "{extra[formatted_data]}\n{exception}"` something that could be (or maybe it is already 👀 ) documented somewhere?

Well, the problem is that this snippet requires firs to access the `LOGURU_FORMAT` variable. However, this variable is defined in `loguru._defaults` which is **not** supposed to be accessed (note the leading underscore).

Documenting this would mean suggesting bad practice since it uses Loguru's non-public API. :/

> Imagine we have this:
> 
> * project A --> support module
> 
> * project B --> has a dependency on project A
> 
> 
> In project A, we create a custom formatter (say to format `dicts` in this case), but we only intend to log this custom format of `dicts` within project A.
> 
> Wouldn't importing project A into project B also affect the logging of dicts contained within project B?
> 
> If so, I think that would be undesirable behaviour? 🤔

If "project B" depends on "project A", that implies "project A" is used as a library. Libraries are not supposed to add any handler to the `logger`, because it does interfere with the end-user code in "project B". See [Configuring Loguru to be used by a library or an application](https://loguru.readthedocs.io/en/stable/resources/recipes.html#configuring-loguru-to-be-used-by-a-library-or-an-application).

If you only want dictionaries from "project A" to be specifically formatted, you need to use your `formatter` within "project B". You can implement a custom `format` discriminating the source:

```python
from project_a import dicts_formatter

def formatter(record):
    if record["name"] == "project_a" or record["name"].startswith("project_a."):
        return dicts_formatter(record)  # Format for "project A"
    return "{time} {message}\n{exception}"  # Format for "project B"
```

Alternatively, you can just document that "project A" puts custom formatted objects in `record["extra"]["formatted_dicts"]` and use it in the `format` of "project B".

**Comment by eitel13** [comment]
@Delgan thank you kindly! 💪 

> Documenting this would mean suggesting bad practice since it uses Loguru's non-public API. :/

Agreed! 👍 

Perhaps in a later version of the public API, a class of available / common (or even just the default) format can be made available to use 🤷‍♂️ But might be a niche case again 🤣 

> Libraries are not supposed to add any handler to the logger, because it does interfere with the end-user code...
> ...you need to use your formatter within "project B"

Good point! And a great reminder to me 🤓 

> You can implement a custom format discriminating the source

Nice! 
 
 > Alternatively, you can just document... 

I like the first suggestion more ☝️ 

Anyway, I think this thread has dragged on a bit, so I appreciate the back and forth _again_...

Have a great 2024! 💪

**Comment by Red-Eyed** [comment]
Why not to use builtin `pformat`?
`logger.info(pprint.pformat(your_dict))`
